### PR TITLE
Fix handling of client connections in pymemcache cache

### PR DESCRIPTION
**Pull request by josephschorr**
pymemcache is apparently not thread safe, so our reuse of the client
was causing the occasional hang on read. This change has us open a new
connection per request, and then close it once complete. We also make
sure to close the DB connection before making the memcache connection,
in case the memcache connection takes a bit of time

We should investigate switching to the PooledClient library for reusing
connections (safely) once we verify this change works

### Description of Changes

* details about the implementation of the changes
* motivation for the change (broken code, new feature, etc)
* contrast with previous behavior

#### Changes:

* ..
* ..

#### Issue: <link to story or task>

**TESTING** ->

**BREAKING CHANGE** ->

---

## Reviewer Checklist

- [ ] It works!
- [ ] Comments provide sufficient explanations for the next contributor
- [ ] Tests cover changes and corner cases
- [ ] Follows Quay syntax patterns and format

### CODE REVIEW

File: data/cache/impl.py [hunk 1]
```diff
@@ -34,6 +37,20 @@ def retrieve(self, cache_key, loader, should_cache=is_not_none):
         pass
 
 
+class DisconnectWrapper(DataModelCache):
+    """ Wrapper around another data model cache that disconnects from the database before
```
Comment (alecmerdler) [inline data/cache/impl.py:41]: Fix docstring to match our convention:
```python
"""
Some docstring.
"""
```

Comment (josephschorr) [reply data/cache/impl.py]: Done

File: data/cache/impl.py [hunk 2]
```diff
@@ -93,6 +111,16 @@ def retrieve(self, cache_key, loader, should_cache=is_not_none):
 _JSON_TYPE = 2
 
 
+@contextmanager
+def memcache_client(client_builder):
+    try:
+        client = client_builder()
+        yield client
+    finally:
```
Comment (alecmerdler) [inline data/cache/impl.py:119]: Doesn't this have to `yield` at every codepath? If an exception is raised before `yield client`, then you'll get: `TypeError: 'NoneType' object is not an iterator`.

Comment (alecmerdler) [reply data/cache/impl.py:119]: https://amir.rachum.com/blog/2017/03/03/generator-cleanup/

File: data/cache/test/test_cache.py [hunk 3]
```diff
@@ -28,6 +34,9 @@ def test_caching(cache_type):
 
 
 def test_memcache():
+    global DATA
```
Comment (alecmerdler) [inline data/cache/test/test_cache.py:37]: What is this madness

Comment (josephschorr) [reply data/cache/test/test_cache.py:37]: Allows us to reassign the top level variable; its just for texting

File: data/cache/impl.py (hunk 2 above)
Comment (josephschorr) [reply data/cache/impl.py:119]: I'll add one in the finally

Comment (alecmerdler) [reply data/cache/impl.py:121]:
```python 
@contextmanager
def memcached_client(client_builder):
    client = None
    try:
        client = client_builder()
    finally:
        yield client
        if client is not None:
            client.cloes()
```

File: data/cache/test/test_cache.py (hunk 3 above)
Comment (alecmerdler) [reply data/cache/test/test_cache.py:37]: Hmm, not a fan but it's okay if just for testing...
